### PR TITLE
Adding orjson package

### PR DESCRIPTION
**Pull request by EntPyle**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml"
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example)
- [x] Source is from official source
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged)
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details)
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there

### CODE REVIEW

**Comment by nehaljwani** [comment]
Hello @EntPyle , I had packaged orjson a while ago like this:

```
{% set name = "orjson" %}
{% set version = "2.6.0" %}

package:
  name: {{ name|lower }}
  version: {{ version }}

source:
  - url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/{{ name }}-{{ version }}.tar.gz
    sha256: c7193de6bd965fb3ee84c72f2126270ecd2e6b755da4d9c7ea47026f9888d16c
    # Lucky for us, this doesn't require glib >2.12
  - url: https://static.rust-lang.org/dist/2020-03-18/rust-nightly-x86_64-unknown-linux-gnu.tar.gz
    sha256: 4ce97a396a4e07fad9336c23dbc99c3f0c5aa41c10b2db003cd0adaecf76547b
    folder: rust-nightly

build:
  number: 0
  skip: True  # [py<36]
  skip: True  # [win]
  script:
    - rust-nightly/install.sh --verbose --prefix=$SRC_DIR/rust-nightly-install --disable-ldconfig
    - export PATH=$SRC_DIR/rust-nightly-install/bin:$PATH
    - "{{ PYTHON }} -m pip install . -vv"

requirements:
  build:
    - {{ compiler('c') }}
  host:
    - python
    - pip
    - maturin
    - toml
  run:
    - python

test:
  imports:
    - orjson

about:
  home: https://github.com/ijl/orjson
  license: Apache 2.0
  license_family: Apache
  license_file: 
      - LICENSE-APACHE
      - LICENSE-MIT
  summary: 'orjson is a fast, correct JSON library for Python.'
  doc_url: https://github.com/ijl/orjson
  dev_url: https://github.com/ijl/orjson
```

**Comment by nehaljwani** [comment]
Your build is currently failing because of a missing dependency: maturin.

You'll see that I am using rust-nightly in my recipe because orjson wanted to use some features which were not available in the stable release, and errored out with:
```error[E0554]: #![feature] may not be used on the stable release channel```

**Comment by EntPyle** [comment]
@nehaljwani Thanks for the assist. I now have a working windows recipe, but linux and osx are failing. From what I know of conda-forge recipes (very little), it looks like you only built for linux and osx. Any insights? I'd like to make this available across the board. Based on the error messages, I think something is up with Cargo or my c compiler on linux and osx, but I'm really not sure.

**Comment by nehaljwani** [comment]
@EntPyle Since we are using the nightly build and not the rust package, we'll probably need to setup a few environment variables too:

```
if [ "$c_compiler" = gcc ] ; then
    case "$BUILD" in
        x86_64-*) rust_env_arch=X86_64_UNKNOWN_LINUX_GNU ;;
        aarch64-*) rust_env_arch=AARCH64_UNKNOWN_LINUX_GNU ;;
        powerpc64le-*) rust_env_arch=POWERPC64LE_UNKNOWN_LINUX_GNU ;;
        *) echo "unknown BUILD $BUILD" ; exit 1 ;;
    esac
   
    export CARGO_TARGET_${rust_env_arch}_LINKER=\$CC
fi
```

**Comment by EntPyle** [comment]
> 
> @EntPyle Since we are using the nightly build and not the rust package, we'll probably need to setup a few environment variables too:
> 

I did acutally include an environment variable for linux in the manner you detailed after looking at the [rust-feedstock](https://github.com/conda-forge/rust-feedstock/blob/master/recipe/build.sh), but it too just failed. I failed to remove the internal build script.

**Comment by nehaljwani** [comment]
You simply copy pasted the part which generates the activation script. You actually need to set the variable here, and not rely on activation scripts, since they won't be of any use.

**Comment by EntPyle** [comment]
@nehaljwani I see. I changed the build.sh script to set the env variable, per your guidance. Where do I pass the actual build instructions if not in that script? 

None of this is now included in my recipe:
`  script:
    - rust-nightly/install.sh --verbose --prefix=$SRC_DIR/rust-nightly-install --disable-ldconfig
    - export PATH=$SRC_DIR/rust-nightly-install/bin:$PATH
    - maturin build --no-sdist --release --strip --manylinux off
    - {{ PYTHON }} -m pip install . -vv`

**Comment by nehaljwani** [comment]
You'll still need the following in build.sh
```
${SRC_DIR}/rust-nightly/install.sh --verbose --prefix=${SRC_DIR}/rust-nightly-install --disable-ldconfig
export PATH=${SRC_DIR}/rust-nightly-install/bin:$PATH
maturin build --no-sdist --release --strip --manylinux off
"${PYTHON}" -m pip install . -vv
```

**Comment by EntPyle** [comment]
@nehaljwani Thank you for all your help. I'm definitely still learning a lot in this arena. Do you have any interest in being added as a maintainer for this package? If not, that's fine, I just thought I'd ask since you are clearly more capable than I in regards to packaging.

**Comment by nehaljwani** [comment]
You'll need to remove `./install.sh --prefix=$PREFIX` since we are doing it below.

**Comment by EntPyle** [comment]
> 
> 
> You'll need to remove `./install.sh --prefix=$PREFIX` since we are doing it below. You can add me as maintainer.

From the build script? I need to delete this line?
`${SRC_DIR}/rust-nightly/install.sh --verbose --prefix=${SRC_DIR}/rust-nightly-install --disable-ldconfig`

File: recipes/orjson/build.sh [hunk 1]
```diff
@@ -0,0 +1,25 @@
+#!/bin/bash
+
+set -ex
+
+./install.sh --prefix=$PREFIX
```
Comment (nehaljwani) [inline recipes/orjson/build.sh:5]: @EntPyle This line. Line 22 is already doing what we want.

Comment (EntPyle) [reply recipes/orjson/build.sh:5]: Got it.

File: recipes/orjson/build.sh [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+if [ "$c_compiler" = gcc ] ; then
+    case "$BUILD" in
+        x86_64-*) rust_env_arch=X86_64_UNKNOWN_LINUX_GNU ;;
+        aarch64-*) rust_env_arch=AARCH64_UNKNOWN_LINUX_GNU ;;
+        powerpc64le-*) rust_env_arch=POWERPC64LE_UNKNOWN_LINUX_GNU ;;
+        *) echo "unknown BUILD $BUILD" ; exit 1 ;;
+    esac
+
+    export CARGO_TARGET_${rust_env_arch}_LINKER=\$CC
```
Comment (nehaljwani) [inline recipes/orjson/build.sh:17]: My bad, we don't need to escape `$` here.

Comment (EntPyle) [reply recipes/orjson/build.sh:16]: Talking about line 15, right?

Comment (nehaljwani) [reply recipes/orjson/build.sh:16]: No, not this. This is the end of `case`, we need that.

I meant, on line 16, we need to replace `\$CC` with `$CC`

Comment (EntPyle) [reply recipes/orjson/build.sh]: That makes much more sense. Updated

File: recipes/orjson/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+source:
+  - url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/{{ name }}-{{ version }}.tar.gz
+    sha256: 910796567d46223a7f4eaea0cc2cb4d3d3adb7a6bccc5ffa8fa2f95b4296d8d9
+    # Lucky for us, this doesn't require glib >2.12
+  - url: https://static.rust-lang.org/dist/2020-04-26/rust-nightly-x86_64-unknown-linux-gnu.tar.gz  # [linux]
+    sha256: a723247878b5b65ec511c86108b9adb99bd1fa9b78ed87986283b93bf9e50d20  # [linux]
+    folder: rust-nightly
+  - url: https://static.rust-lang.org/dist/2020-04-26/rust-nightly-x86_64-apple-darwin.tar.gz  # [osx]
+    sha256: 9f8d7ee88c596465336297aa45c8f9b8d06480728e57b99bb2bc29371d518758  # [osx]
```
Comment (nehaljwani) [inline recipes/orjson/meta.yaml:16]: This SHA256 probably needs to be updated.

File: recipes/orjson/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    folder: rust-nightly
+  - url: https://static.rust-lang.org/dist/2020-04-26/rust-nightly-x86_64-pc-windows-msvc.tar.gz  # [win64]
+    sha256: c6fd886231faa52552757da7a1d813b66aa02413218343089008751fd1b56df7  # [win64]
+    folder: rust-nightly
```
Comment (nehaljwani) [inline recipes/orjson/meta.yaml:20]: For windows, let's try the following in bld.bat:
```
FOR /F "delims=" %%i in ('cygpath.exe -u "%SRC_DIR%\rust-nightly-install"') DO set "pfx=%%i"
bash %SRC_DIR%\rust-nightly\install.sh --verbose --prefix=%pfx% --disable-ldconfig
set PATH=%SRC_DIR%\rust-nightly-install\bin:%PATH%
maturin build --no-sdist --release --strip --manylinux off
%PYTHON% -m pip install . -vv
```

And add: `- posix  # [win]` to the build requirements.

**Comment by nehaljwani** [comment]
The linux build has passed, so we are making good progress!

File: recipes/orjson/build.sh [hunk 5]
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash
+
+set -ex
+
+# Fun times -- by default, Rust/Cargo tries to link executables on Linux by
+# invoking `cc`. An executable of this name is not necessarily available. By
+# setting a magic environment variable, we can override this default.
```
Comment (nehaljwani) [inline recipes/orjson/build.sh:7]: Let's replace these comments with a link to the rust feedstock.

File: recipes/orjson/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+      - LICENSE-APACHE
+      - LICENSE-MIT
+  summary: 'orjson is a fast, correct JSON library for Python.'
+  doc_url: https://github.com/ijl/orjson
+  dev_url: https://github.com/ijl/orjson
+
+extra:
+  recipe-maintainers:
+    # GitHub IDs for maintainers of the recipe.
+    # Always check with the people listed below if they are OK becoming maintainers of the recipe. (There will be spam!)
```
Comment (nehaljwani) [inline recipes/orjson/meta.yaml:55]: Can remove these two comments.

**Comment by EntPyle** [comment]
Great! I've already updated the osx Sha and the license files. I was just
waiting for the linux run to finish. I'll add the windows items you
mentioned and push it all

File: recipes/orjson/meta.yaml (hunk 6 above)
Comment (EntPyle) [reply recipes/orjson/meta.yaml:55]: Done

File: recipes/orjson/build.sh (hunk 5 above)
Comment (EntPyle) [reply recipes/orjson/build.sh:7]: Done

File: recipes/orjson/meta.yaml (hunk 4 above)
Comment (EntPyle) [reply recipes/orjson/meta.yaml:20]: Implemented. Lets see if it works

**Comment by nehaljwani** [comment]
You don't need to add the licenses files in the recipe. They are automatically picked up from the source.

**Comment by EntPyle** [comment]
> 
> 
> You don't need to add the licenses files in the recipe. They are automatically picked up from the source.

But the windows run failed to pick them up. Is that just because it failed elsewhere and wasn't looking in the right place?

**Comment by EntPyle** [comment]
@nehaljwani osx run failed. Is it possible that maturin is not included in the rust nightly apple install?

File: recipes/orjson/meta.yaml (hunk 4 above)
Comment (EntPyle) [reply recipes/orjson/meta.yaml:20]: Failed. Do we need to change the path from the rust nightly path back to the pip src_dir before doing pip install?

**Comment by EntPyle** [comment]
@nehaljwani I'll be honest, I only truly need the linux recipe. While I'd prefer to figure these issues out, I'd be fine with excluding win and osx from the recipe.

**Comment by nehaljwani** [comment]
Okay, add:
```
skip:  True.  # [osx or win]
```

And we'll get this shipped and then worry about the other platforms later.

**Comment by EntPyle** [comment]
I added the skip command but it doesn't appear to have done anything for the CI tests.

**Comment by nehaljwani** [comment]
The skip will be honored later, during the actual build.

**Pull request merged**